File: programmers/그리디/구명보트.py
from collections import deque

def solution(people, limit):
    answer = 0
    
    people = deque(sorted(people))
    
    #print(people)
    
    while people:
        boat = people.pop()
        while people and boat + people[0] <= limit:
            boat += people.popleft()
        answer += 1
    
    return answer

File: programmers/그리디/구명보트.java
import java.util.*;

class Solution {
    public int solution(int[] people, int limit) {
        Arrays.sort(people);
        int answer = 0;
        
        Deque<Integer> q = new ArrayDeque<>();
        
        for (int i = 0; i < people.length; i++){
            q.add(people[i]);
        }
        
        int boat = 0;
        
        while (!q.isEmpty()){
            boat = q.pollLast();
            while (!q.isEmpty()&&q.peek()+boat <= limit){
                boat += q.poll();
            }
            answer += 1;
            
        }
        return answer;
    }
}